cassandra/cassandra_session.go

```diff
@@ -0,0 +1,150 @@
+package cassandra
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/gocql/gocql"
+	log "github.com/sirupsen/logrus"
+)
+
+// Session stores a connection to Cassandra along with associated configurations
+type Session struct {
+	wg                      sync.WaitGroup
+	session                 *gocql.Session
+	cluster                 *gocql.ClusterConfig
+	shutdown                chan struct{}
+	connectionCheckTimeout  time.Duration
+	connectionCheckInterval time.Duration
+	addrs                   string
+	logPrefix               string
+	sync.RWMutex
+}
+
+// NewSession creates and returns a Session. Upon failure it will return nil and an error.
+func NewSession(clusterConfig *gocql.ClusterConfig,
+	timeout time.Duration,
+	interval time.Duration,
+	addrs string,
+	logPrefix string) (*Session, error) {
+	if clusterConfig == nil {
+		log.Errorf("cassandra.NewSession received nil pointer for ClusterConfig")
+		return nil, fmt.Errorf("cassandra.NewSession received nil pointer for ClusterConfig")
+	}
+
+	session, err := clusterConfig.CreateSession()
+	if err != nil {
+		log.Errorf("cassandra.NewSession failed to create session: %v", err)
+		return nil, err
+	}
+
+	cs := &Session{
+		session:                 session,
+		cluster:                 clusterConfig,
+		shutdown:                make(chan struct{}),
+		connectionCheckTimeout:  timeout,
+		connectionCheckInterval: interval,
+		addrs:                   addrs,
+		logPrefix:               logPrefix,
+	}
+
+	if cs.connectionCheckInterval > 0 {
+		cs.wg.Add(1)
+		go cs.deadConnectionRefresh()
+	}
+
+	return cs, nil
+
+}
+
+func (s *Session) Stop() {
+	close(s.shutdown)
+	s.wg.Wait()
+}
+
+// deadConnectionRefresh will run a query using the current Cassandra session every connectionCheckInterval
+// if it cannot query Cassandra for longer than connectionCheckTimeout it will create a new session
+func (s *Session) deadConnectionRefresh() {
```

Comment:
Might make sense to include a comment pointing to gocql/gocql/issues/831 noting that this can go away once that is resolved